working/extension_structs/overview.md

```diff
@@ -0,0 +1,870 @@
+  assert(c1.hashCode == c2.hashCode); // With correspondind hashCode
+
+  print(c1.debugToString()); // Prints "Component(0, 0, 0, 1)"
+
+  print(c1.toString()); // Prints "struct"
+
+  // Generated copy method
+  var c2 = c1.copyWith(x : 2);
+  assert(c2.x == 2);
+}
+```
+
+Structs may explicitly define constructors, static members, and normal class
+members (except fields), and may implement interfaces.  All fields must be
+declared in the header (the "primary constructor"), and are implicitly final
+(making all structs shallowly immutable).  Fields may be marked as private as
+usual by naming them with a leading `_` in their name.
+
+```dart
+struct Component(int _x, int _y, int _z, int _w = 1)
+  implements Comparable<Component> {
+
+  double get x => _x/_w;
+  double get y => _x/_w;
+  double get z => _z/_w;
+
+  int compareTo(Component other) => throw "TODO: Unimplemented";
+
+  Component.zero() : _x = 0, _y = 0, _z = 0;
+}
+```
+
+With the addition of pattern matching and switches, structs will support closed
+families (algebraic datatypes).
+
+```dart
+abstract struct Operand;
+
+struct ConstantOperand(Value c) extends Operand;
+
+struct IdentifierOperand(Identifier i) extends Operand;
+
+Operand replaceIn(Operand oper, Identifier t, Value c) {
+  return switch (oper) {
+    case ConstantOperand(_) : oper;
+    case IdentifierOperand(var i) where i == t: ConstantOperand(c);
+    case IdentifierOperand(_) : oper;
+  }
+}
+```
+
+## Structs in more detail
+
+### Introduction form
+
+We add a new keyword "struct", which is used in place of "class".  The class
+name (plus generics if applicable) must be followed by a parenthesized list of
+field declarations (more on this below). There may optionally be an "extends"
+clause, and an "implements" list.  No mixins are permitted.  After the header,
+the usual brace delimited list of members may be provided subject to
+restrictions described below.  An empty set of members may be elided in favor of
+a semi-colon.  The simplest possible struct definitions then would look like:
+
+```dart
+struct Data();
+struct GenericData<T>();
+```
+
+#### Alternatives
+
+
+##### Re-use the class keyword
+
+An alternative is to continue to use "class" and to use some other piece of
+syntax to indicate that the object in question is not a general class.  For
+example, the parenthesized list might be enough:
+
+
+```dart
+class Data();
+class GenericData<T>();
+```
+
+Alternatively, an additional keyword could be used:
+
+```dart
+class Data wraps ();
+class GenericData<T> wraps ();
+```
+
+Other keywords could be considered.
+
+##### Choose a different leading keyword
+
+Instead of "struct", we could use "view", or some other choice:
+
+```dart
+view Data();
+view GenericData<T>();
+```
+
+Perhaps record:
+
+```dart
+record Data();
+record GenericData<T>();
+```
+
+##### Use a modifier on classes
+
+
+We could use "data class":
+
+```dart
+data class Data();
+data class GenericData<T>();
```

Comment:
@sigmundch had a similar proposal, but using `static data class` instead of your `struct` proposal.  I like his proposal a little better - `struct` on its own doesn't mean much to me for the `view` semantics.  I'm very open to alternative syntax here though.

Reply:
Filed #2363 to track the syntax discussion.